subsys/power/device.c

```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
 /* Number of devices successfully suspended. */
 static device_idx_t num_susp;
 
+static bool should_suspend(const struct device *dev, uint32_t state)
+{
+	int rc;
+	uint32_t current_state;
+
+	rc = device_get_power_state(dev, &current_state);
+	if ((rc != -ENOTSUP) && (rc != 0)) {
+		LOG_DBG("Was not possible to get device %s state: %d",
+			dev->name, rc);
+		return true;
+	}
+
+	/*
+	 * If the device is currently powered off or the request was
+	 * to go to the same state, just ignore it.
+	 */
+	if ((current_state == DEVICE_PM_OFF_STATE) ||
+			(current_state == state)) {
+		return false;
+	}
+
+	return true;
```

Comment:
nit:

could be replaced by `return ((current_state != DEVICE_PM_OFF_STATE) && (current_state != state))`